File: 14-TypeScript/Grey Multi-Tenant/packages/grey-clojure/src/grey/sdk/domain/mutation.clj
(ns grey.sdk.domain.mutation
  "Grey SDK Mutation domain.
  
  Provides mutation operations for modifying data at arbitrary endpoints."
  (:require [grey.sdk.grpc.channel :as channel]
            [grey.sdk.grpc.mutation :as grpc-mutation]
            [grey.sdk.result :as r]
            [grey.sdk.error :as error]
            [clojure.string :as str]))

(def ^:private valid-methods
  #{:post :put :patch :delete})

(defn- validate-mutation
  "Validates mutation parameters."
  [endpoint opts]
  (let [method (get opts :method :post)]
    (cond
      (str/blank? endpoint)
      (r/err (error/validation-error "Endpoint cannot be empty"))
      
      (not (valid-methods method))
      (r/err (error/validation-error (str "Method must be one of: " (str/join ", " (map name valid-methods)))))
      
      :else
      (r/ok {:endpoint endpoint :opts opts}))))

(defn mutate
  "Executes a mutation to the specified endpoint.
  
  Arguments:
    grey-channel - The Grey channel
    endpoint     - Mutation endpoint path
    opts         - Optional parameters:
                   :method  - HTTP method (:post, :put, :patch, :delete) (default: :post)
                   :body    - Request body (any serializable value)
                   :headers - Additional headers (map)
    
  Returns:
    Result with mutation data (:success, :data, :metadata) or error"
  ([grey-channel endpoint]
   (mutate grey-channel endpoint {}))
  ([grey-channel endpoint opts]
   (if (channel/authenticated? grey-channel)
     (r/then (validate-mutation endpoint opts)
             (fn [{:keys [endpoint opts]}]
               (grpc-mutation/mutate grey-channel
                                     (channel/auth-metadata grey-channel)
                                     endpoint
                                     opts)))
     (r/err (error/unauthorized "User not authenticated")))))

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-elixir/lib/grey_sdk/domain/mutation.ex
defmodule GreySdk.Domain.Mutation do
  @moduledoc """
  Domain module for generic mutation operations.

  Provides a flexible mutation interface for custom endpoints.
  """

  alias GreySdk.Error
  alias GreySdk.Grpc.Channel
  alias GreySdk.Grpc.MutationService
  alias GreySdk.Grpc.Types

  @valid_methods ~w(POST PUT PATCH DELETE)

  @doc """
  Executes a mutation against the specified endpoint.

  ## Parameters

  - `channel_server` - The channel GenServer pid
  - `endpoint` - The mutation endpoint path
  - `opts` - Additional options
    - `:method` - HTTP method equivalent (default: "POST")
    - `:body` - Request body

  ## Returns

  - `{:ok, mutation_data}` on success
  - `{:error, %GreySdk.Error{}}` if not authenticated or on failure

  ## Examples

      case GreySdk.Domain.Mutation.mutate(channel, "/api/items", method: "POST", body: %{name: "Item"}) do
        {:ok, %{success: true, data: data}} ->
          IO.puts("Created: \#{inspect(data)}")

        {:ok, %{success: false}} ->
          IO.puts("Mutation failed")

        {:error, error} ->
          IO.puts("Error: \#{error.message}")
      end
  """
  @spec mutate(GenServer.server(), String.t(), keyword()) ::
          {:ok, Types.mutation_data()} | {:error, Error.t()}
  def mutate(channel_server, endpoint, opts \\ []) do
    method = opts |> Keyword.get(:method, "POST") |> String.upcase()

    with :ok <- validate_endpoint(endpoint),
         :ok <- validate_method(method),
         :ok <- validate_authenticated(channel_server),
         {:ok, mutation_data} <- MutationService.mutate(channel_server, endpoint, opts) do
      {:ok, mutation_data}
    else
      {:error, %Error{} = error} -> {:error, error}
      {:error, reason} -> {:error, Error.from_any(reason)}
    end
  end

  # Validation helpers

  defp validate_endpoint(nil), do: {:error, Error.validation("Endpoint is required")}
  defp validate_endpoint(""), do: {:error, Error.validation("Endpoint is required")}
  defp validate_endpoint(endpoint) when is_binary(endpoint), do: :ok
  defp validate_endpoint(_), do: {:error, Error.validation("Endpoint must be a string")}

  defp validate_method(method) when method in @valid_methods, do: :ok

  defp validate_method(_method) do
    {:error, Error.validation("Method must be one of: #{Enum.join(@valid_methods, ", ")}")}
  end

  defp validate_authenticated(channel_server) do
    if Channel.authenticated?(channel_server) do
      :ok
    else
      {:error, Error.unauthorized()}
    end
  end
end
